### DISCUSSION TITLE
Internal traffic to pods shown as Passthrough Cluster in Graph view

### DISCUSSION THREAD

**Discussion by MugenTwo** [kiali/kiali#4454] [Q&A]
Traffic/Edges that goes directly to K8s pods are seemingly being classified as Passthrough cluster.

I was able to reproduce this by using Spring Boot Admin Server, which sends requests (e.g.health checks) to pods (and vice-versa).

1. One of the edges goes from K8s Deployment to the K8s Service of Spring Boot Admin Server. -> This one is correct.
![a](https://user-images.githubusercontent.com/17546545/138561520-e4c391d3-a347-4747-bd8d-edfea0c69fc7.png)

2. Another edge goes from the K8s Deployment of the Spring Boot Admin Server to the K8s Service. -> This one is also correct.
![b](https://user-images.githubusercontent.com/17546545/138561552-7e770d4f-8ea3-4118-8388-03241839ae59.png)

3. And finally, another edge goes from the K8s Deployment of the Spring Boot Admin Server to a Passthrough Cluster (yellow edge) and then to the K8s Deployment. -> This one is incorrect.
![c](https://user-images.githubusercontent.com/17546545/138561616-0ba4aa65-e4f4-4ca9-b0fd-fd557e824d3e.png)

As a consequence, whenever I filter out Spring Admin Server, the Passthrough Cluster of the Spring Admin Server still remains pointing towards the other services which makes it hard to understand what is going on in the system because there are a lot of edges in the graph (as the Spring Boot Admin server pretty much interacts with all the services).

Maybe this is an intended functionality? If it is, is there perhaps way of being able to filter out the traffic to pods (not all Passthrough clusters)? Maybe I have something misconfigured?

Thanks in advanced!

**Comment by jmazzitelli** [comment]
Moving this to "Discussion".

First, review the FAQ talking about unexpected passthrough clusters: https://kiali.io/docs/faq/graph/#why-do-i-see-traffic-to-passthroughcluster

When you see graphs that are unexpected, the first thing you should do is check the Istio telemetry itself - Kiali is merely graphing what is in telemetry. What is usually thought of as a Kiali bug is typically just the way the Istio telemetry is (as found in Prometheus). For details on "How do I inspect the underlying metrics used to generate the Kiali Graph?" see: https://kiali.io/docs/faq/graph/#queryprom

Finally, there is a way to hide certain nodes and edges from the Kiali graph - look at the "hide/find" text boxes on the top of the graph. There is on-screen help available for syntax of the hide/find expressions.

**Comment by MugenTwo** [comment]
Hi @jmazzitelli ! Thanks for taking the time to reply. I believe there is more to discuss regarding this topic.

When the **destination_workload** is correct (like in the image below - where destination_workload="o-services"). Should Istio be sending **destination_service_name** as PassthroughCluster?
![kiali-a](https://user-images.githubusercontent.com/17546545/138683736-507475fa-6887-447c-a8e8-9121f9612649.png)

The thing is pod to pod communication is important for a lot of use cases such as client-side load balancing (e.g. Ribbon), or the case that I have presented (Spring Boot Admin: Health Checks) and I believe that it'll be interesting to address this.

So if the answer to my question is **no**, then I will file a **bug on Istio's repo** to send the correct destination_service_name.
And if the answer is **yes**, then maybe I should file a **Feature request in Kiali** to show this kind of scenarios well. (e.g. by maybe drawing the graph taking into account the destination_workload?)

**Reply by jmazzitelli** [reply]
This is a question for the Istio telemetry folks - but perhaps @jshaughn might know as well.

From what l know (I am no expert on istio telemetry) that seems odd that the destination_service_name is PassthroughCluster when you have destination_workload set to a non-unknown name. It is odd because "PassthroughCluster" doesn't seem to make sense in this use-case (that is, there is no service that the request is being directed to - you are sending requests directly to a pod endpoint circumventing any k8s Service or Istio VirtualService).

I would suggest you ask this question to the Istio folks and see what their expectations are - what do they think the telemetry should look like when not sending requests to a service but instead going directly to a pod (I would not be surprised to hear them say that requests "in the mesh" should go through services and that is why the telemetry looks the way it does - but I'm just guessing).

**Reply by jshaughn** [reply]
I think it's a good question, I would have expected the destination_workload to also be "PassthroughCluster" when the destination_service_name is "PassthroughCluster".  Ask Istio if this is expected.  Also, which version of Istio and Kiali are you using.

**Reply by MugenTwo** [reply]
I've created an issue in Istio here: https://github.com/istio/istio/issues/35726

**Reply by MugenTwo** [reply]
```
➜  ~ istioctl version --short
client version: 1.11.4
control plane version: 1.10.3
data plane version: 1.10.3 (93 proxies)
```
Kiali version: 1.34

**Reply by jshaughn** [reply]
I'm not sure it makes a difference for your situation, but as an FYI, we recommend a minimum Kiali version of v1.38 for use with istio 1.11. 

https://kiali.io/docs/installation/installation-guide/prerequisites/

**Reply by MugenTwo** [reply]
Hi @jshaughn @jmazzitelli ! I've discussed this with people from Istio and they told me that that sending destination_service_name="PassthroughCluster" with destination_workload="o-services" is an intended behaviour: https://github.com/istio/istio/issues/35726#issuecomment-963561863. Which in my opinion actually makes sense because there really is no "service" but there is a "workload" (and app).

So, I think this could be better addressed in Kiali. Maybe with a new arrow color? this arrow could be drawn based on the destination_workload instead of the destination_service_name. Because seeing it as PassthroughCluster or as Unknown in the graph is not useful.

**Reply by jmazzitelli** [reply]
Submit an [enhancement request in the kiali repo](https://github.com/kiali/kiali/issues/new?assignees=&labels=enhancement&template=feature_request.md&title=) and we'll take a look at implementing support for that configuration.
